README.md
# FlashTalent
Change your talents in a flash in World of Warcraft.

## Description
Spend less time fiddling with talents and specializations by changing them in a flash!

FlashTalent is a simple talents addon to make switching talents and specializations them faster and easier. The addon also includes an optional automatic equipment set changer to switch your item set when you change specialization.

### Usage
To begin with you must assign a key bind for the addon. In case one has not been set the addon should prompt for it. If not, go to the Key Bindings menu from the Game Menu (hint: press ESC) and then browse to AddOns category. From there you will find the option to bind a key. Additionally you can also bind key for the quick spec switch and equipment set menu.

If you wish you can also use slash commands **/ft** or **/flashtalent** to open FlashTalent window.

To move the frame hold down alt key and drag it with mouse.

When FlashTalent is open you can click a talent on a row to learn it. This is possible when you are out of combat, and resting or have used required tome or codex.

In the upper right corner you can find the specializations and equipment sets button.

Hovering the button will open specializations menu where you can change your specialization at will to any other. FlashTalent will track what specializations you have used previously and clicking the button after will allow you to quick switch between your most recently used specializations. Optional keybind can also be used for this.

Right click on the button will open equipment sets list where you can equip and update your existing sets and configure automatic set switcher.

## Dependencies
FlashTalent uses Ace3, LibDataBroker, LibQTip and LibSharedMedia which are included in the /libs directory.


CHANGELOG.md
## 4.0.1
* More patch 9.0.1 contributions by Shraen (https://github.com/Shraen):
  * Added Shadowlands Tomes & Codices
  * Auto suggest cheapest Tomes & Codices to use

## 4.0.0
* Patch 9.0.1 fixes contributed by Shraen (https://github.com/Shraen).

## 3.1.0
* Patch 8.1.0 fixes.

## 3.0.0
* Updated for Battle for Azeroth.
* Removed old PVP talent frames.
* Added support for new PVP talent system.
* Added War Mode toggle button.

## 2.2.1
* Restored option to choose if you wish to match specialization names to set names for automatic switching.
* Fixed another equipment set bug.

## 2.2.0
* Updated equipment set features required by game patch changes.
  * Addon now utilizes the native assigned specializations for sets. This also means your previous settings are now reset, sorry. You can reassign sets per spec as before or in Blizzard's equipment set menu.
  * Same change also disables the ability to toggle automatic set changes. Appropriate sets are now always equipped when changing specs.

## 2.1.7
* TOC bump for 7.2.0.

## 2.1.6
* Fixed string formatting error caused by changed WoW API function return values.
* Clamped data broker tooltips to stay within screen area.

## 2.1.5
* Fixed error with honor talents after having reached prestige rank 1 or higher.

## 2.1.4
* TOC bump for 7.1.0.
* Minor fixes.

## 2.1.3
* Fixed automatic tome consume for PvP talents even when setting them should be free.
* Fixed problem where using Tome of the Clear Mind between levels 101 and 109 wouldn't be recognised properly.
* Fixed chat linking locked talents.

## 2.1.2
* Added option to right click in specialization tooltip to switch current quick switch specialization.
* Fixed Tome of the Clear Mind usage level.

## 2.1.1
* You can now link talents directly from FlashTalent by shift-clicking a talent.
* Added notice on the talent tooltip when you have no tomes for auto-use.
* Fixed tooltip note about talent tomes showing up when inappropriate.
* Fixed auto equipper equipping item sets when changing talents if it is enabled.
* Prevent auto tome use if talent row is free.

## 2.1.0
* The key binding to open equipment sets menu at cursor now also opens specializations menu at cursor.
* Added an option to automatically consume a Tome of the Clear/Tranquil Mind when it is necessary. The option is disabled by default.
* Addon should now properly recognise battleground, arena and dungeon preparation grace period.

## 2.0.3
* Fixed nil error with missing equipment set icons.

## 2.0.2
* Removed erroneous tooltip message for honor tab.

## 2.0.1
* Fix to scripts alert caused by the addon.
* Disabled honor tab until level 110.
* Added slash commands **/ft [tab]** and **/flashtalent [tab]**. Command will open FlashTalent window on the given tab index and defaults to class talents tab.

## 2.0.0
* Legion update.
  * Updated to support changed talent API.
  * Added support for Honor talents.
  * You can change between class and honor talents using new tabs next to the talents.
  * Removed glyph support since Legion glyphs are only minor and have mostly one-time use cases. Rest in peace.
  * Changed specializations tooltips to allow changing to any available specialization by clicking them on the list.
    * To counter the removal of dual spec, FlashTalent now tracks what specializations you used last and will allow you to quick change back to the previously used specialization via the specialization button and the optional keybind.
    * Added support for hunter pet specializations.
  * Updated reagent listing to display number of the tomes and the codices instead.
    * Clicking the number will allow you to use a tome or a codex to be able to change talents in non-rest zones.
  * Improved visual look and distinction of empty talent rows.
  * Added new keybind option to directly open honor talents tab. The old keybind now directly opens class talents tab.
  * Added option to change window scale between 80% and 150%.
  * Added option to disable Blizzard alert about unspent talent points.
  * Reorganized internal frame structure so that the whole window stays clamped inside game screen.
  * Disabled challenge mode restriction (temporarily?) because systems have changed and I'm unable to test it without beta.
  * Added a default font changeable in the options menu.
  * Plenty of miscellaneous bug fixes and other minor changes.

## 1.2.0
* Added options menu with various options:
  * Always displaying tooltips when hovering talents and glyphs.
  * Keep window open at all times, even while in combat.
  * Changing glyph buttons anchor to left or right side.
* Added cooldown display for talent rows that displays time until another talent change on the row is possible.
* Switched dual spec change to middle button on the databroker module. Left click now toggles the FlashTalent window.


glyphs.lua
------------------------------------------------------------
-- FlashTalent by Sonaza
-- http://sonaza.com
------------------------------------------------------------

local ADDON_NAME, SHARED = ...;
local _;

local _G = getfenv(0);

local A = unpack(SHARED);

local glyphSlotLevels = {
	[1] = 25, -- Minor 1
	[2] = 25, -- Major 1
	[3] = 50, -- Minor 2
	[4] = 50, -- Major 2
	[5] = 75, -- Minor 3
	[6] = 75, -- Major 3
}

function A:UpdateGlyphs()
	if(true or InCombatLockdown()) then return end
	
	for glyphIndex = 1, 6 do
		local enabled, glyphType, glyphTooltipIndex, glyphSpell, icon, glyphID = GetGlyphSocketInfo(glyphIndex);
		
		local glyph = _G['FlashGlyphsFrameGlyph' .. glyphIndex];
		
		glyph.glyphType = glyphType;
		glyph.ring:SetVertexColor(0, 0, 0);
		
		glyph.levelText:SetText(glyphSlotLevels[glyphIndex]);
		if(not enabled) then
			glyph.levelText:Show();
		else
			glyph.levelText:Hide();
		end
		
		if(glyphID and not A:HasChallengeModeRestriction()) then
			glyph:SetAttribute("shift-type2", "macro");
			glyph:SetAttribute("macrotext",
				"/stopmacro [combat]\n" ..
				"/click PlayerTalentFrameTab3\n"..
				"/click [spec:1] PlayerSpecTab1\n"..
				"/click [spec:2] PlayerSpecTab2\n"..
				"/click GlyphFrameGlyph" .. glyphIndex .. " RightButton\n" ..
				"/click StaticPopup1Button1\n"
			);
		elseif(A:HasChallengeModeRestriction()) then
			glyph:SetAttribute("macrotext", "");
		end
		
		glyph.unlocked = enabled;
		
		if(not enabled) then
			glyph.spell = nil;
			glyph.glyphID = nil;
			
			SetPortraitToTexture(glyph.icon, "Interface\\Icons\\inv_glyph_majordruid");
			glyph.icon:SetVertexColor(0.2, 0.2, 0.2);
			glyph.icon:SetDesaturated(true);
			glyph:SetAlpha(0.9);
		elseif(not glyphSpell) then
			glyph.spell = nil;
			glyph.glyphID = nil;
			
			SetPortraitToTexture(glyph.icon, "Interface\\Icons\\inv_glyph_majordruid");
			glyph.icon:SetVertexColor(0.4, 0.4, 0.4);
			glyph.icon:SetDesaturated(true);
			glyph:SetAlpha(1.0);
		else
			glyph.spell = glyphSpell;
			glyph.glyphID = glyphID;
			
			if(icon) then
				SetPortraitToTexture(glyph.icon, icon);
				glyph.icon:SetVertexColor(1.0, 1.0, 1.0);
				glyph.icon:SetDesaturated(false);
			else
				SetPortraitToTexture(glyph.icon, "Interface\\Icons\\inv_glyph_majordruid");
				glyph.icon:SetVertexColor(0.4, 0.4, 0.4);
				glyph.icon:SetDesaturated(true);
			end
			
			glyph:SetAlpha(1.0);
		end
	end
end

function FlashGlyphButtonTemplate_OnLoad(self)
	self.glow:Play();
	self:RegisterForClicks("LeftButtonUp", "RightButtonUp");
end

function FlashGlyphButtonTemplate_OnEnter(self)
	if(true) then return; end
	
	if(self.unlocked and not A:HasChallengeModeRestriction()) then
		self.glow:Play();
		self.highlight:Show();
		-- self.ring:SetVertexColor(1, 1, 1);
	end
	
	if(IsShiftKeyDown() or A.db.global.AlwaysShowTooltip) then
		local glyphIndex = self:GetID();
		local enabled, glyphType, glyphTooltipIndex, glyphSpell, icon, glyphID = GetGlyphSocketInfo(glyphIndex);
		
		GameTooltip:SetOwner(self, "ANCHOR_RIGHT");
		if(glyphID) then
			GameTooltip:SetGlyphByID(glyphID);
		else
			if(glyphType == 1) then
				GameTooltip:AddLine("Empty Major Glyph Slot");
			elseif(glyphType == 2) then
				GameTooltip:AddLine("Empty Minor Glyph Slot");
			end
		end
		
		if(enabled) then
			if(glyphID) then
				GameTooltip:AddLine("Left click to change the glyph.", 0, 1, 0);
				GameTooltip:AddLine("Shift-Right click to unlearn.", 0, 1, 0);
			else
				GameTooltip:AddLine("Left click to set the glyph.", 0, 1, 0);
			end
		else
			GameTooltip:AddLine(string.format("Glyph Slot Unlocked at Level %d.", glyphSlotLevels[glyphIndex]), 1, 0, 0);
		end
		
		GameTooltip:Show();
	end
	
	A.HoveredGlyph = self;
end

function FlashGlyphButtonTemplate_OnLeave(self)
	GameTooltip:Hide();
	A.HoveredGlyph = nil;
	
	if(not self.isSelected) then
		self.highlight:Hide();
		self.ring:SetVertexColor(0, 0, 0);
	end
end

function FlashGlyphButtonTemplate_PostClick(self, button)
	if(A:HasChallengeModeRestriction()) then
		UIErrorsFrame:AddMessage("Cannot change glyphs while in Challenge Mode.", 1.0, 0.1, 0.1, 1.0);
		return;
	elseif(button == "LeftButton") then
		local glyphIndex = self:GetID();
		local enabled, glyphType, glyphTooltipIndex, glyphSpell, icon, glyphID = GetGlyphSocketInfo(glyphIndex);
		
		if(enabled) then
			if(not self.isSelected) then
				A:ClearSelections(glyphIndex);
				A:OpenGlyphChangeMenu(self, glyphIndex, glyphType, glyphID);
				self.isSelected = true;
			
				self.ring:SetVertexColor(1, 1, 1);
			else
				FlashGlyphChangeFrame:Hide();
				self.highlight:Show();
				self.isSelected = false;
			
				self.ring:SetVertexColor(0, 0, 0);
			end
		end
	end
end

function A:USE_GLYPH()
	if(InCombatLockdown()) then return end
	
	A:ClearSelections();
	FlashGlyphChangeFrame:Hide();
	
	A:UpdateDatabrokerText();
end

function A:GLYPH_UPDATED()
	if(InCombatLockdown()) then return end
	
	A:UpdateGlyphs();
	
	A:UpdateDatabrokerText();
end

function FlashGlyphChangeFrame_OnShow()
	
end

function FlashGlyphChangeFrame_OnHide()
	A:ClearSelections();
end

function A:GetUsableGlyphs()
	local glyphs = {
		[GLYPH_TYPE_MAJOR] = {},
		[GLYPH_TYPE_MINOR] = {},
	};
	
	local numGlyphs = GetNumGlyphs();
	for index = 1, numGlyphs do
		local name, glyphType, isKnown, icon, glyphID, glyphLink, spec, specMatches, excluded = GetGlyphInfo(index);
		if(name ~= "header" and isKnown == true and icon and specMatches) then
			tinsert(glyphs[glyphType], {
				index = index,
				name = name,
				icon = icon,
				glyphID = glyphID,
				spec = spec,
				excluded = excluded,
			});
		end
	end
	
	return glyphs;
end

function A:IsGlyphInUse(glyphID)
	local numGlyphs = GetNumGlyphs();
	for glyphIndex = 1, numGlyphs do
		local enabled, glyphType, glyphTooltipIndex, glyphSpell, icon, socketedGlyphID = GetGlyphSocketInfo(glyphIndex);
		if(glyphID == socketedGlyphID) then return true end
	end
	
	return false;
end

function A:OpenGlyphChangeMenu(glyphFrame, glyphIndex, glyphType, isActive)
	if(InCombatLockdown()) then return end
	if(not (glyphFrame and glyphIndex and glyphType)) then return end
	
	local glyphs = A:GetUsableGlyphs();
	local numGlyphs = #glyphs[glyphType];
	
	local slotName;
	if(glyphType == 1) then
		slotName = string.format("major%d", glyphIndex / 2);
	elseif(glyphType == 2) then
		slotName = string.format("minor%d", (glyphIndex + 1) / 2);
	end
	
	local columns = math.min(6, numGlyphs);
	local rows = math.ceil(numGlyphs / 6);
	
	FlashGlyphChangeFrame:SetWidth(columns * 28);
	FlashGlyphChangeFrame:SetHeight(rows * 28);
	
	local rowFirstButton = FlashGlyphChangeFrameButton1;
	local previousButton = FlashGlyphChangeFrameButton1;
	
	for index = 1, 50 do
		local button = _G['FlashGlyphChangeFrameButton' .. index];
		if(button) then
			button.glyphID = nil;
			button:SetAttribute("type", nil);
			button:Hide()
		else
			break;
		end
	end
	
	if(numGlyphs > 0) then
		local _, reagentCount = GetGlyphClearInfo();
		local hasReagents = (reagentCount or 0) > 0;
		
		FlashGlyphChangeFrame.noGlyphsWarning:Hide();
		
		for index = 1, numGlyphs do
			local data = glyphs[glyphType][index];
			
			local button = _G['FlashGlyphChangeFrameButton' .. index];
			if(not button) then
				button = CreateFrame("Button", 'FlashGlyphChangeFrameButton' .. index, FlashGlyphChangeFrame, "FlashGlyphChangeButtonTemplate");
				
				if(index ~= 1 and (index - 1) % 6 == 0) then
					button:SetPoint("TOPLEFT", rowFirstButton, "BOTTOMLEFT", 0, 0);
				elseif(index ~= 1) then
					button:SetPoint("TOPLEFT", previousButton, "TOPRIGHT", 0, 0);
				end
			end
			
			if(index ~= 1 and (index - 1) % 6 == 0) then
				rowFirstButton = button;
			end
			previousButton = button;
			
			button.freeChange = (glyphFrame.glyphID == nil);
			
			button.glyphID = data.glyphID;
			button.isActive = (data.glyphID == isActive);
			button.glyphInUse = A:IsGlyphInUse(data.glyphID);
			button.excluded = data.excluded;
			
			button.icon:SetTexture(data.icon);
			button.icon:SetVertexColor(0.9, 0.9, 0.9);
			button.icon:SetDesaturated(false);
			
			if(data.excluded) then
				button:SetAttribute("type", nil);
				button.icon:SetVertexColor(1.0, 0.2, 0.2);
				button.icon:SetDesaturated(true);
			elseif(button.glyphInUse) then
				button:SetAttribute("type", nil);
				button.icon:SetVertexColor(0.42, 0.42, 0.42);
				button.icon:SetDesaturated(true);
			elseif(not button.glyphInUse) then
				button:SetAttribute("type", "glyph");
				button:SetAttribute("glyph", data.name);
				button:SetAttribute("slot", slotName);
			end
			
			if(not hasReagents and glyphFrame.glyphID ~= nil) then
				button:SetAttribute("type", nil);
			end
			
			button:Show();
		end
	else
		FlashGlyphChangeFrame.noGlyphsWarning:Show();
		FlashGlyphChangeFrame:SetWidth(160);
		FlashGlyphChangeFrame:SetHeight(32);
	end
	
	FlashGlyphChangeFrame:SetParent(glyphFrame);
	FlashGlyphChangeFrame:ClearAllPoints();
	
	if(self.db.global.AnchorGlyphs == "RIGHT") then
		FlashGlyphChangeFrame:SetPoint("TOPLEFT", glyphFrame, "TOPRIGHT", 8, -1);
	elseif(self.db.global.AnchorGlyphs == "LEFT") then
		FlashGlyphChangeFrame:SetPoint("TOPRIGHT", glyphFrame, "TOPLEFT", -8, -1);
	end
	FlashGlyphChangeFrame:Show();
end

function FlashGlyphChangeButtonTemplate_PostClick(self)
	if(not self.excluded and not self.glyphInUse) then
		local reagent, reagentCount, reagentIcon, _, cost = GetGlyphClearInfo();
			
		if(reagentCount > 0 or self.freeChange or cost == 0) then
			FlashGlyphChangeFrame:Hide();
		elseif(reagentCount == 0 and not self.freeChange) then
			StaticPopup_Show("FLASHTALENT_NOT_ENOUGH_REAGENTS", "glyph", string.format("%s %s", ICON_PATTERN:format(reagentIcon), reagent));
		end
	end
end

function FlashGlyphChangeButtonTemplate_OnEnter(self)
	self.icon:SetVertexColor(1.0, 1.0, 1.0);
	
	if(self.excluded) then
		self.icon:SetVertexColor(1.0, 0.4, 0.4);
	elseif(self.glyphInUse) then
		self.icon:SetVertexColor(0.8, 0.8, 0.8);
	end
	
	GameTooltip:SetOwner(self, "ANCHOR_RIGHT");
	GameTooltip:SetGlyphByID(self.glyphID);
	if(not self.excluded and not self.glyphInUse) then
		GameTooltip:AddLine("Left click to change to this glyph.", 0, 1, 0);
	elseif(self.glyphInUse) then
		GameTooltip:AddLine(" ");
		GameTooltip:AddLine("This glyph is currently in use.", 0.38, 0.77, 1.0);
	end
	GameTooltip:Show();
end

function FlashGlyphChangeButtonTemplate_OnLeave(self)
	self.icon:SetVertexColor(0.9, 0.9, 0.9);
	
	if(self.excluded) then
		self.icon:SetVertexColor(0.9, 0.3, 0.3);
	elseif(self.glyphInUse) then
		self.icon:SetVertexColor(0.3, 0.3, 0.3);
	end
	
	GameTooltip:Hide();
end
	

globals.lua
------------------------------------------------------------
-- FlashTalent by Sonaza
-- All rights reserved
-- http://sonaza.com
------------------------------------------------------------

local ADDON_NAME, Addon = ...;
local _;

FLASHTALENT_ICON_PATTERN              = "|T%s:14:14:0:0|t";
FLASHTALENT_ICON_PATTERN_NOBORDER     = "|T%s:14:14:0:0:64:64:6:58:6:58|t";
SHOW_SPEC_LEVEL			 			  = 7
NEW_CLASS_TALENT_LEVELS 	    	  = { 15, 25, 30, 35, 40, 45, 50};
FLASHTALENT_ICON_ROLES = {
	DAMAGER = "|TInterface\\LFGFRAME\\LFGROLE:14:14:0:0:64:16:16:32:0:16|t",
	TANK    = "|TInterface\\LFGFRAME\\LFGROLE:14:14:0:0:64:16:32:48:0:16|t",
	HEALER  = "|TInterface\\LFGFRAME\\LFGROLE:14:14:0:0:64:16:48:64:0:16|t",
};

BINDING_HEADER_FLASHTALENT = "FlashTalent";
_G["BINDING_NAME_CLICK FlashTalentFrameToggler:LeftButton"] = "Toggle FlashTalent Talents";
_G["BINDING_NAME_FLASHTALENT_CHANGE_DUALSPEC"] = "Quick Switch to Previous Spec";
_G["BINDING_NAME_FLASHTALENT_OPEN_ITEM_SETS_MENU"] = "Open Menus at Cursor";

StaticPopupDialogs["FLASHTALENT_NO_KEYBIND"] = {
	text = "FlashTalent does not currently have a keybinding. Do you want to open the key binding menu to set it?|n|nOption you are looking for is found under AddOns category.",
	button1 = YES,
	button2 = NO,
	button3 = "Don't Ask Again",
	OnAccept = function(self)
		KeyBindingFrame_LoadUI();
		KeyBindingFrame.mode = 1;
		ShowUIPanel(KeyBindingFrame);
	end,
	OnCancel = function(self)
	end,
	OnAlt = function()
		Addon:SetAskedBinding(true);
	end,
	hideOnEscape = 1,
	timeout = 0,
};

StaticPopupDialogs["FLASHTALENT_NOT_ENOUGH_REAGENTS"] = {
	text = "Oops! In order to change the %s you require %s but you have none.",
	button1 = "Okay",
	OnAccept = function(self)
	end,
	hideOnEscape = 1,
	timeout = 0,
};

StaticPopupDialogs["FLASHTALENT_RENAME_EQUIPMENT_SET"] = {
	text = "Enter new name for %s:",
	button1 = SAVE,
	button2 = CANCEL,
	OnAccept = function(self, data)
		local setID = data.setID;
		local oldName = data.oldName;
		local newName = strtrim(self.editBox:GetText());
		
		if(setID and oldName and newName and oldName ~= newName and strlen(newName) > 0) then
			C_EquipmentSet.ModifyEquipmentSet(setID, newName);
		end
	end,
	EditBoxOnEnterPressed = function(self, data)
		local setID = data.setID;
		local oldName = data.oldName;
		local newName = strtrim(self:GetParent().editBox:GetText());
		
		if(setID and oldName and newName and oldName ~= newName and strlen(newName) > 0) then
			C_EquipmentSet.ModifyEquipmentSet(setID, newName);
		end
		
		self:GetParent():Hide();
	end,
	EditBoxOnEscapePressed = function(self)
		self:GetParent():Hide();
	end,
	OnCancel = function(self, data)
		
	end,
	OnShow = function(self, data)
		self.editBox:SetText(data.oldName);
	end,
	OnHide = function(self, data)
		ChatEdit_FocusActiveWindow();
		self.editBox:SetText("");
	end,
	hideOnEscape = 1,
	hasEditBox = 1,
	whileDead = 1,
	timeout = 0,
};


settings.lua
------------------------------------------------------------
-- FlashTalent by Sonaza
-- All rights reserved
-- http://sonaza.com
------------------------------------------------------------

local ADDON_NAME, Addon = ...;
local _;

local AceDB = LibStub("AceDB-3.0");
local LibSharedMedia = LibStub("LibSharedMedia-3.0");

-- Adding default media to LibSharedMedia in case they're not already added
LibSharedMedia:Register("font", "DorisPP", [[Interface\AddOns\FlashTalent\media\DORISPP.TTF]]);

local defaults = {
	char = {
		AskedKeybind        = false,
		AutoSwitchGearSet   = false,
		MatchSpecNames      = nil,
		OpenTalentTab       = 1,
		PreviousSpec        = 0,
	},
	global = {
		AskedKeybind        = false,
		
		FontFace            = "DorisPP",
		Position = {
			Point           = "CENTER",
			RelativePoint   = "CENTER",
			x               = 180,
			y               = 0,
		},
		
		StickyWindow        = false,
		WindowScale         = 1.0,
		IsWindowOpen        = false,
		
		AlwaysShowTooltip   = false,
		AnchorSide          = "RIGHT",
		
		UseReagents         = false,
		
		HideBlizzAlert      = false,
		
	},
};

function Addon:OnInitialize()
	self.db = AceDB:New("FlashTalentDB", defaults);
	
	Addon.CurrentTalentTab = self.db.char.OpenTalentTab;
	
	if(self.db.char.MatchSpecNames == nil) then
		self.db.char.MatchSpecNames = self.db.char.AutoSwitchGearSet;
	end
end

function Addon:UpdateFonts()
	local fontPath = LibSharedMedia:Fetch("font", self.db.global.FontFace);
	FlashTalent_NumberFont_Large_Shadow:SetFont(fontPath, 17, "OUTLINE");
	FlashTalent_NumberFont_Med:SetFont(fontPath, 13, "OUTLINE");
	FlashTalent_NumberFont_Med_Shadow:SetFont(fontPath, 13, "OUTLINE");
	FlashTalent_NumberFont_Small:SetFont(fontPath, 10, "OUTLINE");
end

function Addon:ToggleEscapeClose()
	if(not self.db.global.StickyWindow) then
		tinsert(UISpecialFrames, "FlashTalentFrame");
	else
		for k, v in ipairs(UISpecialFrames) do
			if(v == "FlashTalentFrame") then
				table.remove(UISpecialFrames, k);
				break;
			end
		end
	end
end

function Addon:GetWindowScaleMenu()
	local windowScales = { 0.8, 0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.2, 1.3, 1.4, 1.5, };
	local menu = {};
	
	for index, scale in ipairs(windowScales) do
		tinsert(menu, {
			text = string.format("%d%%", scale * 100),
			func = function() self.db.global.WindowScale = scale; Addon:UpdateFrame(); CloseMenus(); end,
			checked = function() return self.db.global.WindowScale == scale end,
		});
	end
	
	return menu;
end

function Addon:GetSharedFonts()
	local fonts = {};
	for index, font in ipairs(LibSharedMedia:List("font")) do
		tinsert(fonts, {
			text = font,
			func = function()
				self.db.global.FontFace = font;
				Addon:UpdateFrame();
				CloseMenus();
			end,
			checked = function() return self.db.global.FontFace == font; end,
		});
	end
	
	return fonts;
end

function Addon:GetMenuData()
	local data = {
		{
			text = "FlashTalent Options", isTitle = true, notCheckable = true,
		},
		{
			text = "Always show tooltips",
			func = function() self.db.global.AlwaysShowTooltip = not self.db.global.AlwaysShowTooltip; end,
			checked = function() return self.db.global.AlwaysShowTooltip; end,
			isNotRadio = true,
		},
		{
			text = "Keep window open",
			func = function() self.db.global.StickyWindow = not self.db.global.StickyWindow; Addon:ToggleEscapeClose(); end,
			checked = function() return self.db.global.StickyWindow; end,
			isNotRadio = true,
		},
		{
			text = "Hide Blizzard alert about unspent points",
			func = function() self.db.global.HideBlizzAlert = not self.db.global.HideBlizzAlert; end,
			checked = function() return self.db.global.HideBlizzAlert; end,
			isNotRadio = true,
		},
		{
			text = "Automatically use a Tome if necessary",
			func = function() self.db.global.UseReagents = not self.db.global.UseReagents; Addon:UpdateTalentFrame(); end,
			checked = function() return self.db.global.UseReagents; end,
			isNotRadio = true,
		},
		{
			text = " ", isTitle = true, notCheckable = true,
		},
		{
			text = "Miscellaneous", isTitle = true, notCheckable = true,
		},
		{
			text = string.format("|cffffd200Window scale:|r %d%%", Addon.db.global.WindowScale * 100),
			hasArrow = true,
			notCheckable = true,
			menuList = Addon:GetWindowScaleMenu(),
		},
		{
			text = string.format("|cffffd200Font:|r %s", Addon.db.global.FontFace),
			hasArrow = true,
			notCheckable = true,
			menuList = Addon:GetSharedFonts(),
		},
		{
			text = string.format("|cffffd200Anchor side:|r %s", string.lower(self.db.global.AnchorSide)),
			hasArrow = true,
			notCheckable = true,
			menuList = {
				{
					text = "Anchor side", isTitle = true, notCheckable = true,
				},
				{
					text = "To the right",
					func = function() self.db.global.AnchorSide = "RIGHT"; Addon:UpdateFrame(); CloseMenus(); end,
					checked = function() return self.db.global.AnchorSide == "RIGHT"; end,
				},
				{
					text = "To the left",
					func = function() self.db.global.AnchorSide = "LEFT"; Addon:UpdateFrame(); CloseMenus(); end,
					checked = function() return self.db.global.AnchorSide == "LEFT"; end,
				},
			},
		},
		{
			text = " ", isTitle = true, notCheckable = true,
		},
		{
			text = "Close FlashTalent window",
			func = function() FlashTalentFrame:Hide(); end,
			notCheckable = true,
		},
	};
	
	return data;
end

local FlashTalentMenuCursorAnchor;
function Addon:OpenMenusAtCursor()
	if(not FlashTalentMenuCursorAnchor) then
		FlashTalentMenuCursorAnchor = CreateFrame("Frame", "FlashTalentMenuCursorAnchor", UIParent);
		FlashTalentMenuCursorAnchor:SetSize(20, 20);
	end
	
	local x, y = GetCursorPosition();
	local uiscale = UIParent:GetEffectiveScale();
	
	FlashTalentMenuCursorAnchor:ClearAllPoints();
	FlashTalentMenuCursorAnchor:SetPoint("CENTER", UIParent, "BOTTOMLEFT", x / uiscale, y / uiscale);
	
	Addon:OpenItemSetsMenuAtCursor(FlashTalentMenuCursorAnchor);
	Addon:OpenSpecializationsMenuAtCursor(FlashTalentMenuCursorAnchor);
end

function Addon:OpenContextMenu(parentframe)
	if(not Addon.ContextMenu) then
		Addon.ContextMenu = CreateFrame("Frame", "FlashTalentContextMenuFrame", parentframe, "UIDropDownMenuTemplate");
	end
	
	FlashPvpTalentChangeFrame:Hide();
	
	Addon.ContextMenu:SetPoint("BOTTOM", parentframe, "CENTER", 0, 5);
	EasyMenu(Addon:GetMenuData(), Addon.ContextMenu, "cursor", 0, 0, "MENU", 5);
	
	DropDownList1:ClearAllPoints();
	
	if(self.db.global.AnchorSide == "RIGHT") then
		DropDownList1:SetPoint("TOPLEFT", parentframe, "TOPRIGHT", 1, 0);
	elseif(self.db.global.AnchorSide == "LEFT") then
		DropDownList1:SetPoint("TOPRIGHT", parentframe, "TOPLEFT", 1, 0);
	end
	
	DropDownList1:SetClampedToScreen(true);
end

function FlashTalentFrameSettingsButton_OnEnter(self)
	if(DropDownList1:IsVisible()) then return end
	
	GameTooltip:ClearAllPoints();
	GameTooltip:SetOwner(self, "ANCHOR_PRESERVE");
	
	if(Addon.db.global.AnchorSide == "RIGHT") then
		GameTooltip:SetPoint("BOTTOMLEFT", self, "BOTTOMRIGHT", 2, 0);
	elseif(Addon.db.global.AnchorSide == "LEFT") then
		GameTooltip:SetPoint("BOTTOMRIGHT", self, "BOTTOMLEFT", -2, 0);
	end
	
	GameTooltip:AddLine("FlashTalent Pro Tips")
	GameTooltip:AddLine("Switch talents by hovering the one you wish to change to and |cff00ff00left click|r it.", 1, 1, 1, true);
	GameTooltip:AddLine(" ");
	GameTooltip:AddLine("You can switch between class and honor talents by clicking the circles next to talent rows.", 1, 1, 1, true);
	GameTooltip:AddLine(" ");
	GameTooltip:AddLine("The spec switch button will now change between your previously used spec so you can still use it to quickly switch between your most used specializations.", 1, 1, 1, true);
	GameTooltip:AddLine(" ");
	GameTooltip:AddLine("Hold alt and drag to move FlashTalent.", 0, 1, 0);
	if(not Addon.db.global.AlwaysShowTooltip) then
		GameTooltip:AddLine("Hold shift when hovering to display tooltip.", 0, 1, 0);
	end
	GameTooltip:AddLine(" ");
	GameTooltip:AddLine("Click this icon to view options.", 1, 0.8, 0);
	
	GameTooltip:Show();
end

function FlashTalentFrameSettingsButton_OnLeave(self)
	GameTooltip:Hide();
end

function FlashTalentFrameSettingsButton_OnClick(self, button)
	GameTooltip:Hide();
	Addon:OpenContextMenu(self);
end
